### ISSUE TITLE
UTF-8 BOM Handling

### ISSUE THREAD

**Issue by Silvenga** [parcel-bundler/lightningcss#338]
After upgrading Parcel on a project, we started seeing these errors:

```
#26 1.250 @parcel/transformer-css: Unexpected token AtKeyword("import")
#26 1.250
#26 1.250   /source/src/blah/css/app.css:1:2
#26 1.250   > 1 | ﻿@import url('./bootstrap.css');
#26 1.250   >   |  ^
#26 1.250     2 |
#26 1.250     3 | @import url('./materialdesignicons.css');
```

The change in behavior appears to stem from moving over to `lightningcss`.

Looking deeper, it appears the error comes from the UTF-8 BOM at the start of this file.

---

I'm not sure what the CSS spec says about the BOM, but I would propose improving the error message/detection around the BOM to improve the developer experience.

**Comment by richardkmichael** [comment]
@Silvenga

Thank you for mentioning this!  What version are you upgrading from?  I believe it has been happening for awhile.  I tried moving an app to Parcel 2.6 back in August, and had this error.

To repro, I used vim (`:set bomb` / `:set nobomb`) to toggle the BOM on a file in a tiny test project.  The BOM is definitely the problem.

```
# index.js
import './main.css'


# main.css, this file will have a BOM added/removed
@import 'another.css';


# another.css
// any rule
```

```shell

 $ od -t x main.css # BOM --> ..bfbbef
0000000 40bfbbef 6f706d69 27207472 746f6e61
0000020 2e726568 27737363 00000a3b
0000032

 $ npx parcel build --dist-dir ../dist --no-cache index.js
🚨 Build failed.

@parcel/transformer-css: Unexpected end of input

  /user/proj/src/main.css:2:1
    1 | ﻿@import 'another.css';
  > 2 |
  >   | ^


 $ od -t x main.css  # No BOM
0000000 706d6940 2074726f 6f6e6127 72656874
0000020 7373632e 000a3b27
0000027

 $ npx parcel build index.js
✨ Built in 402ms

../index.js     96 B    20ms
../index.css    76 B    20ms
```

Unfortunately, in my case, these are CSS files from a vendor.  I would need to maintain a private fork of their npm package to remove the BOM from all their files.  :-(

Hopefully lightningcss can be changed.

**Comment by peterjanes** [comment]
I don't think this is the right way to handle the problem, but perhaps it might be useful to someone. (I wrote it before realizing https://github.com/parcel-bundler/lightningcss/issues/82 would prevent using it with `bundle()`.)

```
const { transform } = require('lightningcss');

const res = transform({
  filename: 'test.css',
  code: Buffer.from('\ufeff:root { --foo: bar; }'),
  visitor: {
    Selector(selector) {
      if (selector[0].type === 'type' && selector[0].name === '\ufeff') {
        return selector.slice(-1);
      }
      return selector;
    }
  }
});
const code = res.code.toString();
console.log(code, /^:root/.test(code));
```

(For a BOM in CSS in the wild, have a look at [mini.css](https://raw.githubusercontent.com/Chalarangelo/mini.css/master/dist/mini-default.min.css).)

**Comment by DanielVernall** [comment]
I had the same problem when importing a dependency using the @import syntax.
I worked around the issue by saving without the BOM:

VS Code shows the file encoding on the status bar in the bottom right:
![image](https://github.com/parcel-bundler/lightningcss/assets/6403967/b8ff756c-e72f-44d2-b980-e5a79b4bcc93)

You can click the encoding and click "Save with encoding" to changed the file encoding:
![image](https://github.com/parcel-bundler/lightningcss/assets/6403967/3802ba09-9f79-4969-83a8-2eca1f001d8e)

![image](https://github.com/parcel-bundler/lightningcss/assets/6403967/7fbd0847-8e80-4ac3-8415-bbd0ddf59428)

It's interesting that VS Code "guesses" it should be UTF-8 with BOM instead of straight UTF-8.

**Comment by Silvenga** [comment]
There's a couple settings that impact the default encoding - for example any editor configs in the directory tree. For this exact case, it's hard to tell. 

I would expect VSCode to use no BOM by default, unless there's an existing encoding already being used.

**Comment by X-Coder264** [comment]
Still reproducible on 1.33.0, and the blast radius has grown: this no longer only affects
hand-authored files that happen to be saved with a BOM, it now breaks *generated* CSS.

Dart Sass prepends a UTF-8 BOM (instead of `@charset "UTF-8"`) whenever **compressed** output
contains non-ASCII characters — true of any project importing e.g. `bootstrap-sass` glyphicons.
Until postcss 8.5.23 a postcss pass (autoprefixer et al.) silently dropped that BOM; postcss
8.5.24 changed to *"Preserve the BOM after the processing"*, so it now survives to the minifier.
We hit this in a webpack production build (`mini-css-extract-plugin` + `lightningCssMinify`)
right after a routine dependency bump took postcss 8.5.22 → 8.5.25. No file in our source tree
has a BOM anywhere; it is introduced by the compiler and preserved by the transformer.

Repro with the Node API on 1.33.0:

```js
const { transform } = require('lightningcss');
const BOM = '\uFEFF';

const t = (label, css) => {
  try {
    const { code } = transform({ filename: 'in.css', code: Buffer.from(css), minify: true });
    console.log('OK  ', label, '->', JSON.stringify(code.toString()));
  } catch (e) {
    console.log('ERR ', label, '->', e.message, e.loc);
  }
};

t('BOM + rule',            BOM + 'html{color:red}');
t('BOM + comment + rule',  BOM + '/*! banner */html{color:red}');
t('BOM + @charset + rule', BOM + '@charset "UTF-8";html{color:red}');
```

```
OK   BOM + rule            -> "﻿html{color:red}"
ERR  BOM + comment + rule  -> Unexpected token Ident("html") { line: 1, column: 2 }
ERR  BOM + @charset + rule -> Unexpected token AtKeyword("charset") { line: 1, column: 2 }
```

The first line is the one I'd flag as worse than the error message: U+FEFF is a valid ident code
point, so `BOM + html{…}` parses without complaint into a `﻿html` type selector that matches
nothing — the rule is silently dropped from the page, no error, no warning. When a comment or
at-rule follows the BOM, the same lexing yields two idents separated by non-whitespace, which is
where the parse error above comes from. That is also why the failure looks so mysterious in a
bundler: the reported location is `1:2` of a generated file, and the identifier named in the
message is the first selector of the first stylesheet, dozens of KB away from the actual cause.

Per [CSS Syntax §3.2](https://www.w3.org/TR/css-syntax-3/#input-byte-stream) the BOM is supposed
to be removed in the decode step, before tokenising, so both behaviours look like conformance
bugs rather than only DX issues.

On the performance concern raised earlier in this thread: this is a single 3-byte prefix check
per stylesheet at parse entry (`code.strip_prefix('\u{feff}')`), not per token, so it shouldn't
be measurable. Happy to send a PR — the only thing I'd want a maintainer's call on first is
whether `bundle()` should strip it from every imported file as well as from the entry.
